### ISSUE TITLE
UA_OATD类forward函数疑惑

### ISSUE THREAD

**Issue by dadqad** [ChenhaoWang2000/UA-OATD#1]
您好，想想您请教几个代码中的问题。
这是有疑问处代码，uaoatd.py UA_OATD类中forward函数
`    def forward(self, trajs, lengths, batch_size, mode, c):
        adj = array2sparse(self.adj)        #加载邻接矩阵 A
        d_norm = array2sparse(self.d_norm)  # 加载 D^(-1/2)
        H = self.Norm_A(adj, d_norm)        # D^(-1/2)*A*D(-1/2) 计算出H

        e_nodes = self.embedding(self.nodes)    # 0-token_size 
        e_nodes = H.mm(e_nodes).mm(self.V)      # H*(We*e(g))*(V)
        e_input = torch.index_select(e_nodes, 0, trajs.reshape(1, -1).squeeze().to(args.device)). \
            reshape(batch_size, -1, args.embedding_size) #根据索引取得对应的邻接嵌入
        d_input = torch.cat(
            (torch.zeros(batch_size, 1, args.embedding_size, dtype=torch.long).to(args.device), e_input[:, :-1, :]),
            dim=1)

        decoder_inputs = pack_padded_sequence(d_input, lengths, batch_first=True, enforce_sorted=False) #打包mask`
        
问题1:  `d_input = torch.cat((torch.zeros(batch_size, 1, args.embedding_size, dtype=torch.long).to(args.device), e_input[:, :-1,:]),dim=1)` 此处代码是干什么用的？我认为就是把把其中一个(batch_size,1,embedding_size)的0，拼接到    e_input中，但是e_input又去除了最后(batch_size,127,embedding_size)的数据。这个部分我不能理解您的意图。

问题2:  论文中是Unified embedding = Adjacency Embedding+  Grid Embedding 但是这一串的代码，我只看到了Adjacency Embedding的计算，并未发现Grid Embedding。

问题3: self.V对应论文哪一个权重矩阵？

希望您能为我解答一下我的疑惑，十分谢谢。

**Comment by ChenhaoWang2000** [comment]
问题1:在seq2seq模型中，通常decoder的输入会右移一位，采用在之前时刻的输出作为当前时刻的输入，这种方式可以防止将当前的信息泄露。比如transformer模型中的shifted right操作，和轨迹异常检测的GM-VSAE模型都采用了这种方式，可以参考下面两张图的红色框出部分：
![transformer](https://github.com/ChenhaoWang2000/UA-OATD/assets/54847470/397a8539-142b-4f83-9b6e-55c82539af6e)
![GM-VSAE](https://github.com/ChenhaoWang2000/UA-OATD/assets/54847470/834e233f-4888-4596-835f-9b4d61f0b71e)

问题2，问题3：在计算邻接矩阵A时已经加入了单位矩阵`I`，可以认为单位矩阵对应的是Grid Embedding，邻接矩阵对应的是Adjacency Embedding。self.V对应的是W_g。上传的代码是对一些权重简化过的版本，抱歉给您的理解造成了麻烦。

**Comment by dadqad** [comment]
感谢您的回答。

**Comment by dadqad** [comment]
$def ~ forward():$
$~~~~~~adj  =  (A + I)$
$~~~~~~d_{norm} ={{\rm{D}}^{ - \frac{1}{2}}}$
$~~~~~~{\rm{       H = }}{{\rm{D}}^{ - \frac{1}{2}}}(A + I){{\rm{D}}^{ - \frac{1}{2}}}$
$~~~~~~V = W_g $
$~~~~~~{e\underline{ }nodes} = e(g)$
$~~~~~~{e\underline{ }nodes} =D^{-\frac{1}{2}}(A+I)D^{-\frac{1}{2}}e(g){W_g}=D^{-\frac{1}{2}}AD^{-\frac{1}{2}}e(g){W_g}+D^{-1}e(g){W_g}=Unified embedding$
请问一下我对于您代码中简化后的矩阵公式理解有错误吗？
